### PR TITLE
feat: add ICU MessageFormat support

### PR DESCRIPTION
**Pull request by Syjalo**
**Please describe the changes this PR makes and why it should be merged**
This PR adds support for ICU MessageFormat and replaces all string variables to the format.
Language plural rules: https://unicode-org.github.io/cldr-staging/charts/37/supplemental/language_plural_rules.html

**In this PR:**
- Code changes were not tested
- Strings were edited (text changed while the key stayed the same)
- Strings were added/removed (new/deleted string keys)
- An open issue was fixed/addressed: #696

### CODE REVIEW

File: src/commands/Utility/minecraft.ts [hunk 1]
```diff
@@ -133,14 +133,13 @@ const command: Command = {
...
-									? getString(isOwnUser ? "history.youChangedName1" : "history.userChangedName1", { variables: { username } })
-									: getString(isOwnUser ? "history.youChangedName" : "history.userChangedName", {
-											variables: { username, number: nameHistory.length - 1 },
-									  })
-								: getString(isOwnUser ? "history.youNeverChanged" : "history.userNeverChanged", { variables: { username } }),
+						description: isOwnUser
+							? getString("history.youChangedName", {
+									variables: { number: nameHistory.length - 1 },
+							  })
+							: getString("history.userChangedName", {
```
Comment (ImRodry) [inline src/commands/Utility/minecraft.ts:140]: you can simplify this by putting the ternary in the path parameter instead of two getString calls

File: src/listeners/interactionCreate.ts [hunk 2]
```diff
@@ -136,8 +137,15 @@ client.on("interactionCreate", async interaction => {
...
 					}
-					if (typeof string === "string" && variables)
-						for (const [variable, text] of Object.entries(variables)) string = string.replace(`%%${variable}%%`, String(text))
+					if (typeof string === "string" && variables) {
+						const discordLocale = transformBotLocale(lang) ?? "en-GB"
+						string = new MessageFormat(
+							{
+								[discordLocale]: (value: string | number, ord?: boolean) =>
+									new Intl.PluralRules(discordLocale, { type: ord ? "ordinal" : "cardinal" }).select(Number(value)),
+							}[discordLocale],
```
Comment (ImRodry) [inline src/listeners/interactionCreate.ts:146]: you're creating an object with 1 key just to index it straight after? Why?

File: strings/en/global.json [hunk 3]
```diff
@@ -29,35 +29,35 @@
 		"wrongLink": "Wrong link!"
 	},
 	"roles": {
-		"noLevel": "You must be at least chat level %%level%% to receive this role! You can check your current chat level using %%command%% in %%channel%%.",
-		"successGive": "Successfully gave you the %%role%% role!",
-		"successTake": "Successfully took the %%role%% role from you!"
+		"noLevel": "You must be at least chat level {level} to receive this role! You can check your current chat level using {command} in {channel}.",
```
Comment (ImRodry) [inline strings/en/global.json:32]: yeah uhm can we not do this please

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: src/listeners/interactionCreate.ts (hunk 2 above)
Comment (Syjalo) [reply src/listeners/interactionCreate.ts:146]: > you're creating an object with 1 key just to index it straight after? Why?

Default plural function is throws an error if an unsupported form for locale was used (zero, one, few, etc).
To make a custom plural function you need to name it by the locale. I don't know another way to name it with a variable.

File: strings/en/global.json (hunk 3 above)
Comment (Syjalo) [reply strings/en/global.json]: wdym?

Comment (ImRodry) [reply strings/en/global.json]: I'm not gonna merge this with every single variable changed

File: src/listeners/interactionCreate.ts (hunk 2 above)
Comment (ImRodry) [reply src/listeners/interactionCreate.ts:146]: you can create a function without making it in an object

Comment (Syjalo) [reply src/listeners/interactionCreate.ts:146]: The name of the function should be a locale. How should I take the name of the function from a variable?

File: strings/en/global.json (hunk 3 above)
Comment (Syjalo) [reply strings/en/global.json]: You have access to the branch. You can do as you need.

Comment (ImRodry) [reply strings/en/global.json]: But it seems like you changed the code and I'm not yet familiar with this package so if possible I'd like you to keep the variables that don't need to interact with this package to be kept as they were and for that check to be added in the code

File: strings/en/global.json [hunk 4]
```diff
@@ -29,7 +29,7 @@
 		"wrongLink": "Wrong link!"
 	},
 	"roles": {
-		"noLevel": "You must be at least chat level %%level%% to receive this role! You can check your current chat level using %%command%% in %%channel%%.",
+		"noLevel": "You must be at least chat level {level, number} to receive this role! You can check your current chat level using %%command%% in %%channel%%.",
```
Comment (ImRodry) [inline strings/en/global.json:32]: what does this mean? Why do you need to have `, number`?

File: strings/en/languagestats.json [hunk 5]
```diff
@@ -1,8 +1,8 @@
 {
 	"moduleName": "Language statistics",
 	"statsAll": "Here are the statistics for the %%language%% language on all the projects we currently support.",
-	"translated": "%%percentage%%% translated (%%translated%%/%%total%% strings)",
-	"approved": "%%percentage%%% approved (%%approved%%/%%total%% strings)",
+	"translated": "{percentage, number, :: percent} translated ({translated, number}/{total, number} strings)",
+	"approved": "{percentage, number, :: percent} approved ({approved, number}/{total, number} strings)",
```
Comment (ImRodry) [inline strings/en/languagestats.json:5]:
```suggestion
	"translated": "{percentage, number, :: percent}% translated ({translated, number}/{total, number} strings)",
	"approved": "{percentage, number, :: percent}% approved ({approved, number}/{total, number} strings)",
```

File: strings/en/global.json (hunk 4 above)
Comment (Syjalo) [reply strings/en/global.json:32]: > what does this mean? Why do you need to have `, number`?

Convert the number to local one

en-GB: 1,234.56
ru: 1 234,56
ar: ١٬٢٣٤٫٥٦

Comment (ImRodry) [reply strings/en/global.json:32]: And do you need that there? Can't you make that distinction in the code?

File: strings/en/languagestats.json (hunk 5 above)
Comment (Syjalo) [reply strings/en/languagestats.json:5]: It automatically adds it

File: strings/en/global.json (hunk 4 above)
Comment (Syjalo) [reply strings/en/global.json:32]: > And do you need that there? Can't you make that distinction in the code?

I can. Should I do that?

Comment (ImRodry) [reply strings/en/global.json:32]: Yes. You should avoid all string changes that aren't absolutely necessary

File: strings/en/levels.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
 	"moduleName": "XP Leaderboard",
 	"pageTitle": "Here is the XP leaderboard!",
-	"level": "#%%rank%% | Level %%level%% | XP: %%xp%%",
+	"level": "#%%rank%% | Level %%level%% | XP: {xp, number, :: .00 compact-short}",
```
Comment (ImRodry) [inline strings/en/levels.json:4]: what is this

File: strings/en/prefix.json [hunk 7]
```diff
@@ -16,8 +16,8 @@
 	"caution": "Caution",
 	"warning": "Setting your prefix to a language you don't translate will result in a warning. Do you want to change your nickname anyway?",
 	"react": "React with all flags you want to add to your prefix in order.",
-	"reactTimer": "You have %%cooldown%% seconds. Hit ❎ to cancel.",
-	"reactTimer2": "You have less than %%cooldown%% seconds left. Hit ✅ to confirm or ❎ to cancel.",
+	"reactTimer": "You have {cooldown, plural, one {# second} other {# seconds}}. Hit ❎ to cancel.",
+	"reactTimer2": "You have less than {cooldown, plural, one {# second} other {# seconds}} left. Hit ✅ to confirm or ❎ to cancel.",
```
Comment (ImRodry) [inline strings/en/prefix.json:20]: is this needed? These are fixed values that are either 30 or 60 seconds

File: strings/en/rank.json [hunk 8]
```diff
@@ -5,7 +5,7 @@
 	"youNotRanked": "You're not ranked yet!",
 	"userNotRanked": "That user is not ranked yet!",
 	"howRank": "In order to be ranked you must talk in the chats!",
-	"textProgress": "%%currentXp%%/%%xpNeeded%% | %%messages%% Messages",
+	"textProgress": "{currentXp, number, :: .00 compact-short}/{xpNeeded, number, :: .00 compact-short} | {messages, number, :: .00 compact-short} Messages",
```
Comment (ImRodry) [inline strings/en/rank.json:8]: again, why is this needed

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: strings/en/prefix.json (hunk 7 above)
Comment (Syjalo) [reply strings/en/prefix.json:20]: Some languages may have different words for 30 and 60

Comment (ImRodry) [reply strings/en/prefix.json:20]: What I'm saying is this is likely always the same value, I can double check but I'm pretty sure it is

File: src/listeners/interactionCreate.ts [hunk 9]
```diff
@@ -136,8 +137,16 @@ client.on("interactionCreate", async interaction => {
 								console.error(`Couldn't get string ${path} in English for ${file}, please fix this`)
 						}
 					}
-					if (typeof string === "string" && variables)
+					if (typeof string === "string" && variables) {
 						for (const [variable, text] of Object.entries(variables)) string = string.replace(`%%${variable}%%`, String(text))
+						const discordLocale = transformBotLocale(lang) ?? "en-GB"
```
Comment (ImRodry) [inline src/listeners/interactionCreate.ts:142]: why wouldn't this support all languages?

Comment (Syjalo) [reply src/listeners/interactionCreate.ts:142]: If the locale is `en` then `transformBotLocale` will return `null`

Comment (ImRodry) [reply src/listeners/interactionCreate.ts:142]: That doesn't answer my question. Why doesn't this support all the languages on the bot?

File: src/listeners/interactionCreate.ts [hunk 10]
```diff
@@ -136,8 +137,16 @@ client.on("interactionCreate", async interaction => {
 								console.error(`Couldn't get string ${path} in English for ${file}, please fix this`)
 						}
 					}
-					if (typeof string === "string" && variables)
+					if (typeof string === "string" && variables) {
 						for (const [variable, text] of Object.entries(variables)) string = string.replace(`%%${variable}%%`, String(text))
+						const discordLocale = transformBotLocale(lang)!
```
Comment (ImRodry) [inline src/listeners/interactionCreate.ts:142]: I'm gonna repeat the question: why does this only support the locales Discord supports and not all the ones we support on the bot?

Comment (Syjalo) [reply src/listeners/interactionCreate.ts:142]: The library supports only `xx-XX` format locales

Comment (ImRodry) [reply src/listeners/interactionCreate.ts:142]: Ok? Then transform them

Comment (Syjalo) [reply src/listeners/interactionCreate.ts:142]: > Ok? Then transform them

The bot locales format is `xx_XX`, the library supports `xx-XX`.

File: src/listeners/interactionCreate.ts [hunk 11]
```diff
@@ -136,8 +137,16 @@ client.on("interactionCreate", async interaction => {
 								console.error(`Couldn't get string ${path} in English for ${file}, please fix this`)
 						}
 					}
-					if (typeof string === "string" && variables)
+					if (typeof string === "string" && variables) {
 						for (const [variable, text] of Object.entries(variables)) string = string.replace(`%%${variable}%%`, String(text))
+						const locale = lang.replace("_", "-")
```
Comment (ImRodry) [inline src/listeners/interactionCreate.ts:142]: You can probably call formatNumberToLocaleString here instead of calling it on every single instance where a number is passed as a variable

File: src/commands/Info/levels.ts [hunk 12]
```diff
@@ -111,7 +111,13 @@ function fetchPage(page: number, pages: DbUser[][], getString: GetStringFunction
 		author: { name: getString("moduleName") },
 		title: getString("pageTitle"),
 		footer: {
-			text: getString("pagination.page", { variables: { number: page + 1, total: pages.length }, file: "global" }),
+			text: getString("pagination.page", {
+				variables: {
+					number: formatNumberToLocaleString(page + 1, getString),
+					total: formatNumberToLocaleString(pages.length, getString),
```
Comment (ImRodry) [inline src/commands/Info/levels.ts:117]: why is this needed?

File: src/commands/Info/levels.ts [hunk 13]
```diff
@@ -132,7 +138,7 @@ function fetchPage(page: number, pages: DbUser[][], getString: GetStringFunction
 			})
 		} else {
 			pageEmbed.addFields({
-				name: getString("unranked", { variables: { rank: i + 1 + page * 24 } }),
+				name: getString("unranked", { variables: { rank: formatNumberToLocaleString(i + 1 + page * 24, getString) } }),
```
Comment (ImRodry) [inline src/commands/Info/levels.ts:141]: why is this needed?

File: src/lib/util.ts [hunk 14]
```diff
@@ -167,6 +167,10 @@ export function createModlogEmbed(embedData: EmbedData, modlog: PunishmentLog, m
 	return new EmbedBuilder(embedData)
 }
 
+export function formatNumberToLocaleString(number: number, getString: GetStringFunction) {
+	return number.toLocaleString(getString("region.dateLocale", { file: "global" }))
+}
```
Comment (ImRodry) [inline src/lib/util.ts:172]: IMO you can remove this function and instead change parseToNumberString to use toLocaleString if the number is smaller than 1k

File: src/commands/Info/rank.ts [hunk 15]
```diff
@@ -51,10 +51,13 @@ const command: Command = {
...
 				title: user.id === interaction.user.id ? getString("yourRank") : getString("userRank", { variables: { user: user.tag } }),
-				description:
-					user.id === interaction.user.id
-						? getString("youLevel", { variables: { level: userDb.levels.level, rank: ranking } })
-						: getString("userLevel", { variables: { user: `${user}`, level: userDb.levels.level, rank: ranking } }),
+				description: getString(user.id === interaction.user.id ? "youLevel" : "userLevel", {
+					variables: {
+						user: `${user}`,
+						level: formatNumberToLocaleString(userDb.levels.level, getString),
+						rank: formatNumberToLocaleString(ranking, getString),
```
Comment (ImRodry) [inline src/commands/Info/rank.ts:58]: ditto

File: src/commands/Utility/help.ts [hunk 16]
```diff
@@ -87,7 +87,9 @@ const command: Command = {
 			pages.forEach(page => {
 				if (page.number === 0) return
 				page1.addFields({
-					name: getString("pageNumber", { variables: { number: page.number, total: pages.length } }),
+					name: getString("pageNumber", {
+						variables: { number: formatNumberToLocaleString(page.number, getString) },
```
Comment (ImRodry) [inline src/commands/Utility/help.ts:91]: ditto

File: src/commands/Utility/help.ts [hunk 17]
```diff
@@ -190,8 +192,8 @@ function fetchPage(page: number, pages: Page[], getString: GetStringFunction, in
 				footer: {
 					text: getString("pagination.page", {
 						variables: {
-							number: page + 1,
-							total: pages.length,
+							number: formatNumberToLocaleString(page + 1, getString),
+							total: formatNumberToLocaleString(pages.length, getString),
```
Comment (ImRodry) [inline src/commands/Utility/help.ts:196]: ditto

File: src/commands/Utility/languagestats.ts [hunk 18]
```diff
@@ -84,15 +84,15 @@ const command: Command = {
 				name: "Hypixel",
 				value: `${getString("translated", {
 					variables: {
-						percentage: hypixelData.translationProgress,
-						translated: hypixelData.phrases.translated,
-						total: hypixelData.phrases.total,
+						percentage: formatNumberToLocaleString(hypixelData.translationProgress, getString),
```
Comment (ImRodry) [inline src/commands/Utility/languagestats.ts:87]: ditto (only for this one)

File: src/commands/Utility/languagestats.ts [hunk 19]
```diff
@@ -84,15 +84,15 @@ const command: Command = {
...
+						translated: formatNumberToLocaleString(hypixelData.phrases.translated, getString),
+						total: formatNumberToLocaleString(hypixelData.phrases.total, getString),
 					},
 				})}\n${getString("approved", {
 					variables: {
-						percentage: hypixelData.approvalProgress,
-						approved: hypixelData.phrases.approved,
-						total: hypixelData.phrases.total,
+						percentage: formatNumberToLocaleString(hypixelData.approvalProgress, getString),
```
Comment (ImRodry) [inline src/commands/Utility/languagestats.ts:93]: ditto

File: src/commands/Utility/languagestats.ts [hunk 20]
```diff
@@ -102,15 +102,15 @@ const command: Command = {
 				name: "Quickplay",
 				value: `${getString("translated", {
 					variables: {
-						percentage: quickplayData.translationProgress,
-						translated: quickplayData.phrases.translated,
-						total: quickplayData.phrases.total,
+						percentage: formatNumberToLocaleString(quickplayData.translationProgress, getString),
```
Comment (ImRodry) [inline src/commands/Utility/languagestats.ts:105]: ditto

File: src/commands/Utility/languagestats.ts [hunk 21]
```diff
@@ -102,15 +102,15 @@ const command: Command = {
...
+						translated: formatNumberToLocaleString(quickplayData.phrases.translated, getString),
+						total: formatNumberToLocaleString(quickplayData.phrases.total, getString),
 					},
 				})}\n${getString("approved", {
 					variables: {
-						percentage: quickplayData.approvalProgress,
-						approved: quickplayData.phrases.approved,
-						total: quickplayData.phrases.total,
+						percentage: formatNumberToLocaleString(quickplayData.approvalProgress, getString),
```
Comment (ImRodry) [inline src/commands/Utility/languagestats.ts:111]: ditto

File: src/commands/Utility/languagestats.ts [hunk 22]
```diff
@@ -120,15 +120,15 @@ const command: Command = {
 				name: "SkyblockAddons",
 				value: `${getString("translated", {
 					variables: {
-						percentage: sbaData.translationProgress,
-						translated: sbaData.phrases.translated,
-						total: sbaData.phrases.total,
+						percentage: formatNumberToLocaleString(sbaData.translationProgress, getString),
```
Comment (ImRodry) [inline src/commands/Utility/languagestats.ts:123]: ditto

File: src/commands/Utility/languagestats.ts [hunk 23]
```diff
@@ -120,15 +120,15 @@ const command: Command = {
...
+						translated: formatNumberToLocaleString(sbaData.phrases.translated, getString),
+						total: formatNumberToLocaleString(sbaData.phrases.total, getString),
 					},
 				})}\n${getString("approved", {
 					variables: {
-						percentage: sbaData.approvalProgress,
-						approved: sbaData.phrases.approved,
-						total: sbaData.phrases.total,
+						percentage: formatNumberToLocaleString(sbaData.approvalProgress, getString),
```
Comment (ImRodry) [inline src/commands/Utility/languagestats.ts:129]: ditto

File: src/commands/Utility/languagestats.ts [hunk 24]
```diff
@@ -138,12 +138,16 @@ const command: Command = {
 				name: "Hypixel Translators Bot",
 				value: `${getString("translated", {
 					variables: {
-						percentage: botData.translationProgress,
-						translated: botData.phrases.translated,
-						total: botData.phrases.total,
+						percentage: formatNumberToLocaleString(botData.translationProgress, getString),
```
Comment (ImRodry) [inline src/commands/Utility/languagestats.ts:141]: ditto

File: src/commands/Utility/languagestats.ts [hunk 25]
```diff
@@ -138,12 +138,16 @@ const command: Command = {
...
+						translated: formatNumberToLocaleString(botData.phrases.translated, getString),
+						total: formatNumberToLocaleString(botData.phrases.total, getString),
 					},
 				})}\n${getString("approved", {
-					variables: { percentage: botData.approvalProgress, approved: botData.phrases.approved, total: botData.phrases.total },
+					variables: {
+						percentage: formatNumberToLocaleString(botData.approvalProgress, getString),
```
Comment (ImRodry) [inline src/commands/Utility/languagestats.ts:147]: ditto

File: src/commands/Utility/minecraft.ts [hunk 26]
```diff
@@ -133,20 +133,21 @@ const command: Command = {
...
 						fields: constructFields(pages[page]),
 						footer: {
 							text:
 								pages.length === 1
 									? randomTip
-									: getString("pagination.page", { variables: { number: page + 1, total: pages.length }, file: "global" }),
+									: getString("pagination.page", {
+											variables: {
+												number: formatNumberToLocaleString(page + 1, getString),
+												total: formatNumberToLocaleString(pages.length, getString),
```
Comment (ImRodry) [inline src/commands/Utility/minecraft.ts:147]: ditto

File: src/commands/Utility/poll.ts [hunk 27]
```diff
@@ -309,7 +309,9 @@ const command: Command = {
 						// Make sure to account for NaN values
 						value: `${generateProgressBar(o.votes.length, totalVoteCount)} ${
 							Math.round((o.votes.length / totalVoteCount) * 100) || 0
-						}% (**${getString(o.votes.length === 1 ? "voteCount" : "voteCountPlural", { variables: { number: o.votes.length } })}**)`,
+						}% (**${getString(o.votes.length === 1 ? "voteCount" : "voteCountPlural", {
+							variables: { number: formatNumberToLocaleString(o.votes.length, getString) },
```
Comment (ImRodry) [inline src/commands/Utility/poll.ts:313]: ditto

File: src/commands/Utility/quote.ts [hunk 28]
```diff
@@ -146,7 +146,7 @@ async function findQuote(randomTip: string, interaction: ChatInputCommandInterac
 			color: colors.error,
 			author: { name: getString("moduleName") },
 			title: getString("invalidArg"),
-			description: getString("indexArg", { variables: { arg: quoteId!, max: count } }),
+			description: getString("indexArg", { variables: { arg: quoteId!, max: formatNumberToLocaleString(count, getString) } }),
```
Comment (ImRodry) [inline src/commands/Utility/quote.ts:149]: ditto

File: src/interactions/buttons.ts [hunk 29]
```diff
@@ -33,7 +34,13 @@ export default async function handleButtonInteractions(interaction: ButtonIntera
 			if ((userDb.levels?.level ?? 0) < 5) {
 				console.log(`${interaction.member.user.tag} tried to get the Giveaway pings role but they're level ${userDb.levels?.level ?? 0} lol`)
 				return await interaction.reply({
-					content: getString("roles.noLevel", { variables: { level: 5, command: "`/rank`", channel: `<#${ids.channels.bots}>` } }),
+					content: getString("roles.noLevel", {
+						variables: {
+							level: formatNumberToLocaleString(5, getString),
```
Comment (ImRodry) [inline src/interactions/buttons.ts:39]: ditto

**Review by ImRodry — CHANGES_REQUESTED** [review]
I believe a bunch of these changes with the formatNumberToLocaleString are not needed (not all, however)

Also I think you missed some places where we could be using ICU like in the cooldown system (for second/seconds), could you take another look?

File: src/commands/Info/levels.ts (hunk 12 above)
Comment (Syjalo) [reply src/commands/Info/levels.ts:117]: `formatNumberToLocaleString' formats to a locale string. If there are no changes with small numbers in some languages, it changes in Arabic.

en: 123
ar: ١٢٣

**Comment by Syjalo** [comment]
> Also I think you missed some places where we could be using ICU like in the cooldown system (for second/seconds), could you take another look?

The cooldown system uses Discord timestamp. Should I replace it?

**Comment by ImRodry** [comment]
> > Also I think you missed some places where we could be using ICU like in the cooldown system (for second/seconds), could you take another look?
> 
> The cooldown system uses a Discord timestamp. Should I replace it?

No it doesn’t. It says X second(s)/minutes

File: src/commands/Info/levels.ts (hunk 12 above)
Comment (ImRodry) [reply src/commands/Info/levels.ts]: That change seems out of the scope of this PR and honestly I’m not sure if we’d wanna go down that path.

File: src/commands/Info/ping.ts [hunk 30]
```diff
@@ -28,7 +28,7 @@ const command: Command = {
 			color,
 			author: { name: getString("moduleName") },
 			title: getString("pong", { variables: { pingEmote: "<:ping:620954198493888512>" } }),
-			description: `${getString("message", { variables: { ping: ping } })}\n\n${getString("onlineSince", {
+			description: `${getString("message", { variables: { ping } })}\n\n${getString("onlineSince", {
```
Comment (ImRodry) [inline src/commands/Info/ping.ts:31]: Out of the scope, also we do this for a reason

File: strings/en/mention.json [hunk 31]
```diff
@@ -1,8 +1,4 @@
 {
-	"errorNoPing": "You don't have that language's proofreader role, ",
-	"errorNoPingPr": "so you can't mention it.",
-	"errorNoPingTr": "so you can't mention its translator role.",
-	"errorNoPingAll": "so you can't mention the entire language.",
-	"errorNoPingDisclaimer": "Contact the server owner or administrator if you really need to.",
+	"errorNoPing": "You don't have that language's proofreader role, {role, select, pr {so you can't mention it} tr {so you can't mention its translator role} other {so you can't mention the entire language}}. Contact the server owner or administrator if you really need to.",
```
Comment (ImRodry) [inline strings/en/mention.json:2]: Please adjust the code accordingly
```suggestion
	"errorNoPing": "You don't have that language's proofreader role, {role, select, pr {so you can't mention it} tr {so you can't mention its translator role} all {so you can't mention the entire language}}. Contact a server administrator if you really need to.",
```

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: src/commands/Info/levels.ts (hunk 12 above)
Comment (Syjalo) [reply src/commands/Info/levels.ts]: > That change seems out of the scope of this PR and honestly I’m not sure if we’d wanna go down that path

If you won't format it everywhere it will be formated to locale string in some strings and in others it won't.

File: strings/en/mention.json (hunk 31 above)
Comment (Syjalo) [reply strings/en/mention.json:2]: There's always should be `other`

Comment (ImRodry) [reply strings/en/mention.json]: Edited

File: src/commands/Utility/help.ts [hunk 32]
```diff
@@ -163,11 +163,14 @@ const command: Command = {
 			})
 			if (cmdDesc !== getString("inDev")) {
 				if (cmd.cooldown) {
-					if (cmd.cooldown >= 120)
-						embed.addFields({ name: getString("cooldownField"), value: `${cmd.cooldown / 60} ${getString("minutes")}`, inline: true })
-					else if (cmd.cooldown === 1)
-						embed.addFields({ name: getString("cooldownField"), value: `${cmd.cooldown} ${getString("second")}`, inline: true })
-					else embed.addFields({ name: getString("cooldownField"), value: `${cmd.cooldown} ${getString("seconds")}`, inline: true })
+					if (cmd.cooldown >= 120) {
```
Comment (ImRodry) [inline src/commands/Utility/help.ts:166]: Because you gave `minutes` a plural string too
```suggestion
					if (cmd.cooldown >= 60) {
```

**Review by ImRodry — CHANGES_REQUESTED** [review]
Since we're using 2 variable types now, do you think you can make a function that compares variables between two strings by comparing the arrays created by the regexes of the two variable types? Also please update the getGlobalString function on messageCreate.ts

File: src/lib/util.ts [hunk 33]
```diff
@@ -47,22 +47,29 @@ const activeConnections: string[] = []
...
 // #region Functions
 
-export function arrayEqual(a: unknown, b: unknown) {
-	if (a === b) return true
+export function checkVariables(firstString: string, secondString: string) {
+	const match = (string: string) =>
+			Array.from(string.matchAll(/%%(\w+)%%|{(\w+)(, \w+(, .+)?)?}/g))
+				.map(el => el[1] ?? el[2])
+				.sort(),
+		arrayEqual = (a: unknown, b: unknown) => {
```
Comment (ImRodry) [inline src/lib/util.ts:55]: Please keep the arrayEqual function separate and sort this function alphabetically within the file

**Review by ImRodry — APPROVED** [review]
